### PR TITLE
Make `NonEmptyArrayType::toArray()` return `$this`

### PR DESCRIPTION
**Pull request by zonuexe**
close https://github.com/phpstan/phpstan/issues/9208

### CODE REVIEW

File: tests/PHPStan/Analyser/data/array-slice.php [hunk 1]
```diff
@@ -12,7 +12,7 @@ class Foo
 	 */
 	public function nonEmpty(array $a): void
 	{
-		assertType('array', array_slice($a, 1));
+		assertType('non-empty-array', array_slice($a, 1));
```
Comment (zonuexe) [inline tests/PHPStan/Analyser/data/array-slice.php]: `array_slice()` can return empty arrays. https://3v4l.org/8VGrv

File: src/Type/Php/ArraySliceFunctionReturnTypeExtension.php [hunk 2]
```diff
@@ -50,7 +51,7 @@ public function getTypeFromFunctionCall(FunctionReflection $functionReflection,
 		}
 
 		if ($valueType->isIterableAtLeastOnce()->yes()) {
-			return $valueType->toArray();
+			return TypeCombinator::union($valueType->toArray(), new NonEmptyArrayType());
```
Comment (ondrejmirtes) [inline src/Type/Php/ArraySliceFunctionReturnTypeExtension.php]: intersect, not union

File: tests/PHPStan/Analyser/data/bug-9208.php [hunk 3]
```diff
@@ -0,0 +1,18 @@
+<?php
+
+/**
+ * @param int|non-empty-list<int> $id_or_ids
+ * @return non-empty-list<int>
+ */
+function f(int|array $id_or_ids): array
+{
+	if (is_array($id_or_ids)) {
+		\PHPStan\Testing\assertType('non-empty-list<int>', (array)$id_or_ids);
```
Comment (ondrejmirtes) [inline tests/PHPStan/Analyser/data/bug-9208.php]: The file needs to have namespace `Bug9208`.

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/data/bug-9208.php]: The file isn't referenced in NodeScopeResolverTest

Comment (ondrejmirtes) [reply tests/PHPStan/Analyser/data/bug-9208.php]: Please use `PHPStan\Testing\assertType` in a `use function` statement.

**Review by ondrejmirtes — CHANGES_REQUESTED** [review]

File: src/Type/Php/ArraySliceFunctionReturnTypeExtension.php (hunk 2 above)
Comment (zonuexe) [reply src/Type/Php/ArraySliceFunctionReturnTypeExtension.php]: I'm trying to make a `non-empty-array&array<T>` into an `array<T>`. Isn't it a union?

File: tests/PHPStan/Analyser/data/bug-9208.php (hunk 3 above)
Comment (zonuexe) [reply tests/PHPStan/Analyser/data/bug-9208.php]: fixed

File: src/Type/Php/ArraySliceFunctionReturnTypeExtension.php (hunk 2 above)
Comment (ondrejmirtes) [reply src/Type/Php/ArraySliceFunctionReturnTypeExtension.php]: So you mean that you want an array that’s non-empty to no longer be non-empty? I’m afraid you need to recreate it from scratch by doing new ArrayType.

File: tests/PHPStan/Analyser/data/array-slice.php [hunk 4]
```diff
@@ -9,10 +9,14 @@ class Foo
...
 	/**
 	 * @param non-empty-array $a
+	 * @param non-empty-list $b
+	 * @param non-empty-array<int>|non-empty-list<string> $c
 	 */
-	public function nonEmpty(array $a): void
+	public function nonEmpty(array $a, array $b, array $c): void
 	{
 		assertType('array', array_slice($a, 1));
+		assertType('list', array_slice($b, 1));
```
Comment (staabm) [inline tests/PHPStan/Analyser/data/array-slice.php]: seems this assertion fails atm in CI 
```suggestion
		assertType('list<mixed>', array_slice($b, 1));
```

**Comment by zonuexe** [comment]
@ondrejmirtes Could you please review this PR?

**Comment by herndlm** [comment]
Ah interesting, this is the same problem I found independently via https://github.com/phpstan/phpstan-src/pull/2451#issuecomment-1587862340 but with a different solution that not just only removes that one accessory

**Comment by ondrejmirtes** [comment]
There's a much easier solution to make a non-empty array possibly empty again :) https://github.com/phpstan/phpstan-src/pull/2349/commits/9e84671c475f2e1fded227965852ce9a32788484

**Review by ondrejmirtes — APPROVED** [review]

**Comment by ondrejmirtes** [comment]
Thank you!

**Comment by zonuexe** [comment]
> There's a much easier solution to make a non-empty array possibly empty again :) https://github.com/phpstan/phpstan-src/commit/9e84671c475f2e1fded227965852ce9a32788484

[I agree.](https://github.com/phpstan/phpstan-src/pull/2349#discussion_r1170464977) 😄

**Comment by ondrejmirtes** [comment]
Oh wow, so you actually planted it in my brain two months ago :D Well I'm a bit embarassed now :D

**Pull request merged**